File: react/src/modelDownloader.ts
import RNFS from 'react-native-fs'

const defaultModelFolderName: string = "models"
const defaultModelUrl: string = "https://huggingface.co/unsloth/SmolLM2-135M-Instruct-GGUF/resolve/main/SmolLM2-135M-Instruct-Q8_0.gguf"

const supportedProviders: string[] = ['huggingface.co']
const supportedFormats: string[] = ['gguf']

export class ModelDownloader {
    modelUrl: string
    modelFolderName: string
    modelName: string = ''
    fullModelPath: string = ''
    fullModelFolderPath: string = ''

    constructor(modelUrl?: string, modelFolderName?: string) {
        this.modelUrl = modelUrl || defaultModelUrl
        this.modelFolderName = modelFolderName || defaultModelFolderName
        this.validateURLFormat();
        this.parseModelName();
    }

    validateURLFormat() {
        let isValid: boolean = true
        if (!this.modelUrl.startsWith('http://') && !this.modelUrl.startsWith('https://')) {
            console.log('Invalid model URL', this.modelUrl)
            isValid = false
        }
        if (!supportedProviders.includes(this.modelUrl.split('//').at(1)?.split('/').at(0) || '')) {
            console.log('Invalid model provider', this.modelUrl.split('//').at(0) || '')
            isValid = false
        }
        if (!supportedFormats.includes(this.modelUrl.split('.').at(-1) || '')) {
            console.log('Invalid model format', this.modelUrl.split('.').at(-1) || '')
            isValid = false
        }
        if (!isValid) {
            throw new Error("Invalid model URL")
        }
        console.log('validateURLFormat', isValid)
    }

    parseModelName() {
        const path = this.modelUrl.split('/').filter(Boolean)
        this.modelName = path.pop() || ''
        this.fullModelFolderPath = `${RNFS.DocumentDirectoryPath}/${this.modelFolderName}`
        this.fullModelPath = `${this.fullModelFolderPath}/${this.modelName}`
    }

    async createModelFolder() {
        if (!(await RNFS.exists(this.fullModelFolderPath))) {
            await RNFS.mkdir(this.fullModelFolderPath)
            console.log('Model folder created', this.fullModelFolderPath)
        }
        console.log('model folder exists', this.fullModelFolderPath)
    }

    async downloadModelIfNotExists(
        onProgress?: (progress: number) => void,
        onSuccess?: (modelPath: string) => void
    ): Promise<string> {
        if (!(await RNFS.exists(this.fullModelPath))) {
            await this.createModelFolder()
            console.log('downloading model...')
            await RNFS.downloadFile({
                fromUrl: this.modelUrl,
                toFile: this.fullModelPath,
                begin: (response: RNFS.DownloadBeginCallbackResult) => {console.log('begin', response)}, // if you don't provide begin, progress won't work!!
                // progress: onProgress ? (response: RNFS.DownloadProgressCallbackResult) => {console.log('progress', response)} : ()=>{},
                progress: onProgress ? (response: RNFS.DownloadProgressCallbackResult) => {
                    if (response.contentLength > 0) {
                        let percentage = Math.floor(response.bytesWritten/response.contentLength*100);
                        onProgress(percentage)
                    }
                } : () => {},
                progressInterval: 100,
            }).promise.then(() => {
                onSuccess?.(this.fullModelPath)
            }).catch((error: any) => {
                throw new Error(`Error downloading model: ${error}`)
            })
        }
        return this.fullModelPath;
    }
}

File: flutter/lib/src/model_downloader.dart
import 'dart:io';
import 'package:path_provider/path_provider.dart';
import 'package:dio/dio.dart';
import 'package:path/path.dart' as p;

/// Model downloader utility for downloading and managing LLM models
class ModelDownloader {
  final String modelUrl;
  final String modelFolderName;
  late String modelName;
  late String fullModelPath;
  late String fullModelFolderPath;
  
  final List<String> supportedProviders = ['huggingface.co'];
  final List<String> supportedFormats = ['gguf'];
  
  static const String defaultModelFolderName = "models";
  static const String defaultModelUrl = "https://huggingface.co/unsloth/SmolLM2-135M-Instruct-GGUF/resolve/main/SmolLM2-135M-Instruct-Q8_0.gguf";
  
  /// Create a new ModelDownloader instance
  /// 
  /// [modelUrl] is the URL to download the model from
  /// [modelFolderName] is the name of the folder to store models in
  ModelDownloader({
    String? modelUrl,
    String? modelFolderName,
  }) : 
    modelUrl = modelUrl ?? defaultModelUrl,
    modelFolderName = modelFolderName ?? defaultModelFolderName {
    validateUrlFormat();
    parseModelName();
  }
  
  /// Validate the model URL format
  void validateUrlFormat() {
    bool isValid = true;
    if (!modelUrl.startsWith('http://') && !modelUrl.startsWith('https://')) {
      print('Invalid model URL: $modelUrl');
      isValid = false;
    }
    
    final Uri uri = Uri.parse(modelUrl);
    final String? host = uri.host;
    if (host == null || !supportedProviders.contains(host)) {
      print('Invalid model provider: $host');
      isValid = false;
    }
    
    final String extension = p.extension(modelUrl).toLowerCase().replaceAll('.', '');
    if (!supportedFormats.contains(extension)) {
      print('Invalid model format: $extension');
      isValid = false;
    }
    
    if (!isValid) {
      throw Exception("Invalid model URL");
    }
    
    print('Valid URL format: $modelUrl');
  }
  
  /// Parse the model name from the URL
  Future<void> parseModelName() async {
    final Uri uri = Uri.parse(modelUrl);
    final String path = uri.path;
    modelName = p.basename(path);
    
    final appDocDir = await getApplicationDocumentsDirectory();
    fullModelFolderPath = '${appDocDir.path}/$modelFolderName';
    fullModelPath = '$fullModelFolderPath/$modelName';
    
    print('Model name: $modelName');
    print('Model folder path: $fullModelFolderPath');
    print('Model path: $fullModelPath');
  }
  
  /// Create the model folder if it doesn't exist
  Future<void> createModelFolder() async {
    final dir = Directory(fullModelFolderPath);
    if (!await dir.exists()) {
      await dir.create(recursive: true);
      print('Model folder created: $fullModelFolderPath');
    } else {
      print('Model folder already exists: $fullModelFolderPath');
    }
  }
  
  /// Download the model if it doesn't exist
  /// 
  /// [onProgress] is called with the download progress (0-100)
  /// [onSuccess] is called when the download completes successfully
  Future<String> downloadModelIfNotExists({
    Function(int progress)? onProgress,
    Function(String modelPath)? onSuccess,
  }) async {
    final file = File(fullModelPath);
    
    if (!await file.exists()) {
      await createModelFolder();
      print('Downloading model from $modelUrl');
      
      final dio = Dio();
      
      try {
        await dio.download(
          modelUrl,
          fullModelPath,
          onReceiveProgress: (received, total) {
            if (total != -1) {
              final percentage = ((received / total) * 100).floor();
              onProgress?.call(percentage);
            }
          },
        );
        
        print('Model downloaded successfully to $fullModelPath');
        onSuccess?.call(fullModelPath);
      } catch (e) {
        print('Error downloading model: $e');
        throw Exception('Error downloading model: $e');
      }
    } else {
      print('Model already exists at $fullModelPath');
      onSuccess?.call(fullModelPath);
    }
    
    return fullModelPath;
  }
}

/// Download a model if it doesn't exist
/// 
/// [modelUrl] is the URL to download the model from
/// [modelFolderName] is the name of the folder to store models in
/// [onProgress] is called with the download progress (0-100)
/// [onSuccess] is called when the download completes successfully
Future<String> downloadModelIfNotExists({
  String? modelUrl,
  String? modelFolderName,
  Function(int progress)? onProgress,
  Function(String modelPath)? onSuccess,
}) async {
  final downloader = ModelDownloader(
    modelUrl: modelUrl,
    modelFolderName: modelFolderName,
  );
  
  return downloader.downloadModelIfNotExists(
    onProgress: onProgress,
    onSuccess: onSuccess,
  );
} 